src/tick.rs

```diff
@@ -0,0 +1,453 @@
+use std::str::FromStr;
+
+use cosmwasm_std::Uint128;
+use cosmwasm_std::{Decimal, Decimal256, Storage};
+
+use crate::state::{TickExpIndexData, TICK_EXP_CACHE};
+use crate::ContractError;
+
+const MAX_SPOT_PRICE: &str = "100000000000000000000000000000000000000"; // 10^35
+const MIN_SPOT_PRICE: &str = "0.000000000001"; // 10^-12
+const EXPONENT_AT_PRICE_ONE: i64 = -6;
+pub const MIN_INITIALIZED_TICK: i64 = -108000000;
+pub const MAX_TICK: i128 = 342000000;
+
+// The methods in this file are copied from the Quasar cl vault contract.
```

Comment:
This whole file was taken from Quasar's vault contract. That's not to say it shouldn't be reviewed, but wanted to highlight this.